### PR TITLE
Improve Dictionary Writer Performance

### PR DESCRIPTION
**Pull request by arunthirupathi**
Dictionary Writers are significantly slower than the Direct writer, 
even when taking cost of Dictionary maintenance into account.
This change tries to close the gap. It has variety of improvements.

1. DictionaryBuilder is only used for non null elements. 
Dictionary Writer represents the null by an invalid index(-1).
2. DictionaryBuilder does not expose the Block. It is used
as a pure dictionary. This change opens up the possibility of
using non-sequential buffers to store the dictionary.
3. On abandoning String dictionary, a dictionary block creation
is avoided. Creating dictionary blocks and creating slice
per row, produces too much garbage. Slices with offset and lengths
are directly written to the SliceDirectColumnWriter.
4. Statistics will not be recomputed when abandoning String
dictionary. Existing statistics computed by String dictionary writer
will be reused.
5. RowGroupIndexes are converted from IntBigArray to int[]. IntBigArray
uses 1024 segments (4 KB) and default stride size is 10_000. So only
8 of them could be filled at a time and wasting 4KB per stride.
This on some tests account for 20% of the memory usage.
6. Rehashing of dictionary, need not verify the element equality.
Previously rehashing was verifying the element equals.

There is 20-30% improvement on dictionary encoders with this change.

Test plan -
Added new additional test coverage and enhanced dictionary tests.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* Improve dictionary writer performance
```

### CODE REVIEW

**Comment by arunthirupathi** [comment]
New Performance numbers after the fixes. (The writerDictionaryAndConvert benchmark had a bug for Varchar and 100% uniqueness, substantial improvement is because of that).

| (typeSignature) | (uniqueValuesPercentage) | writeDirect | writeDictionaryAndConvert | writeDictionary |
| --------------- | ------------------------ | ----------- | ------------------------- | --------------- |
| bigint          | 1                        | 100.00%     | 187.44%                   | 106.09%         |
|                 | 5                        | 100.00%     | 216.78%                   | 133.93%         |
|                 | 10                       | 100.00%     | 221.10%                   | 154.43%         |
|                 | 100                      | 100.00%     | 361.99%                   | 309.30%         |
| integer         | 1                        | 100.00%     | 210.08%                   | 198.28%         |
|                 | 5                        | 100.00%     | 230.75%                   | 230.51%         |
|                 | 10                       | 100.00%     | 259.07%                   | 258.10%         |
|                 | 100                      | 100.00%     | 743.09%                   | 793.16%         |
| varchar         | 1                        | 100.00%     | 253.79%                   | 260.03%         |
|                 | 5                        | 100.00%     | 313.32%                   | 422.99%         |
|                 | 10                       | 100.00%     | 397.26%                   | 650.91%         |
|                 | 100                      | 100.00%     | 1348.82%                  | 4872.76%        |

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java [hunk 1]
```diff
@@ -14,38 +14,32 @@
...
 import static it.unimi.dsi.fastutil.HashCommon.arraySize;
 import static java.lang.Math.min;
 import static java.util.Objects.requireNonNull;
 
 // TODO this class is not memory efficient.  We can bypass all of the Presto type and block code
-// since we are only interested in a hash of byte arrays.  The only place an actual block is needed
-// is during conversion to direct, and in that case we can use a slice array block.  This code
-// can use store the data in multiple Slices to avoid a large contiguous allocation.
+// since we are only interested in a hash of byte arrays.
 public class DictionaryBuilder
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java:30]: Given we are for sure specializing this class. Let's rename it to reflect it

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java [hunk 2]
```diff
@@ -71,7 +75,7 @@
     private final List<DictionaryRowGroup> rowGroups = new ArrayList<>();
     private long columnStatisticsRetainedSizeInBytes;
 
-    private IntBigArray rowGroupIndexes;
+    private int[] rowGroupIndexes;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java:78]: Is this safe? We will face gigantic int[] in memory. We used to have the problem for Presto on Spark. cc: @arhimondr

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java [hunk 3]
```diff
@@ -81,42 +70,47 @@ public long getRetainedSizeInBytes()
...
     {
-        return elementBlock;
+        return elementBlock.compareTo(
+                left,
+                0,
+                elementBlock.getSliceLength(left),
+                elementBlock,
+                right,
+                0,
+                elementBlock.getSliceLength(right));
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java:82]: Same comment as the one for the last PR: it is fairly hard to follow all the refactor + new changes within one commit. Let's maybe start with this PR to break it into multiple commits. Some commits should only contain code movement. Some commits should contain interface changes. Some commits have logic changes. Data structure change like BigIntArray to int[] should also be a separate commit.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java (hunk 2 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java:78]: rowGroupIndexes are typically 10K and I haven't seen it configured ever. IntBigArray has 1024 per segment and 1024 segments. Only 10 segments are used today and rest 1014 segments point to null. So per 10K rows, we are wasting 4KB with using IntBigArray there.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java (hunk 3 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java:82]: I will break them into smaller commits and update this PR with a list of commits for easy reviewing.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java (hunk 1 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryBuilder.java:30]: Some names I can think of are SliceDicitonary or SliceDictionaryBuilder. Please let me know if you prefer some other names.

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 4]
```diff
@@ -1680,7 +1680,6 @@ public static OrcSelectiveRecordReader createCustomOrcSelectiveRecordReader(
                 DwrfKeyProvider.of(intermediateEncryptionKeys));
 
         assertEquals(orcReader.getColumnNames().subList(0, types.size()), makeColumnNames(types.size()));
-        assertEquals(orcReader.getFooter().getRowsInRowGroup(), 10_000);
```
Comment (bhhari) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:1683]: why is this check removed?

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java [hunk 5]
```diff
@@ -186,55 +174,29 @@ protected ColumnStatistics createColumnStatistics()
...
+        int[] sortedDictionaryIndexes = getSortedDictionary(dictionary);
         for (int sortedDictionaryIndex : sortedDictionaryIndexes) {
-            if (!dictionaryElements.isNull(sortedDictionaryIndex)) {
-                int length = dictionaryElements.getSliceLength(sortedDictionaryIndex);
-                dictionaryLengthStream.writeLong(length);
-                Slice value = dictionaryElements.getSlice(sortedDictionaryIndex, 0, length);
-                dictionaryDataStream.writeSlice(value);
-            }
+            int length = dictionary.getSliceLength(sortedDictionaryIndex);
+            dictionaryLengthStream.writeLong(length);
```
Comment (bhhari) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java:194]: is this optimal when compared to the previous one?

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java (hunk 4 above)
Comment (arunthirupathi) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:1683]: It blocks testing strides with different rows. I added a new test with stride length of 14,XXX rows to validate that row group conversion works as expected.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java (hunk 5 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java:194]: This avoids creating intermediary slice value (temporary object). rawSlice is reference to the same object and this uses offsets to write directly, so it is more efficient.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java [hunk 6]
```diff
@@ -151,6 +202,22 @@ public void testStringNonRepeatingValues()
         }
     }
 
+    @Test
+    public void testStringIncreasedStrideSize()
+            throws Exception
+    {
+        ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java:209]: `... = ImmutableList.builder();`

File: presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java [hunk 7]
```diff
@@ -523,4 +768,4 @@ void validate(int batchId, OrcWriter writer)
             }
         }
     }
-}
+}
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java]: we need a new line after this.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java [hunk 8]
```diff
@@ -91,10 +91,10 @@ public int getDictionaryEntries()
     }
 
     @Override
-    protected boolean tryConvertRowGroupToDirect(int dictionaryIndexCount, int[] dictionaryIndexes, int maxDirectBytes)
+    protected boolean tryConvertRowGroupToDirect(int dictionaryIndexCount, int[] dictionaryIndexes, int maxDirectBytes, Optional<ColumnStatistics> statistics)
     {
+        directColumnWriter.beginRowGroup();
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java:97]: begin and finishRowGroup here should be orthogonal from stats reuse. Any better abstraction we can think of to avoid the implication?

**Review by highker — COMMENTED** [review]
mostly LGTM. Just one comment on stats reuse

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java (hunk 8 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java:97]: Here is my thinking: previously tryConvertToDirect was responsible for calling the beginRowGroup and finishRowGroup. After this change each rowGroup write itself will be responsible. Presence of statistics means that, rowGroup is complete and finishRowGroup should be called. If statistics does not exist, then it is a partial row group and finishRowGroup should not be called. But beginRowGroup happens unconditionally.

**Comment by highker** [comment]
While I'm thinking how to make the open/finishRowGroup more generic, can help breakdown the last commit "Address code review comments" into each previous commits so that each individual commit is clean?

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java [hunk 9]
```diff
@@ -103,8 +103,14 @@ protected boolean tryConvertRowGroupToDirect(int dictionaryIndexCount, int[] dic
...
-                Slice slice = elementBlock.getSlice(dictionaryIndex, 0, length);
-                directColumnWriter.writeSlice(slice);
+                if (statistics.isPresent()) {
+                    Slice slice = elementBlock.getSlice(dictionaryIndex, 0, length);
+                    directColumnWriter.writeSliceWithoutStatistics(slice, 0, length);
+                }
+                else {
+                    Slice slice = elementBlock.getSlice(dictionaryIndex, 0, length);
+                    directColumnWriter.writeSlice(slice);
+                }
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java]:
```java
Slice slice = elementBlock.getSlice(dictionaryIndex, 0, length);
if (statistics.isPresent()) {
    directColumnWriter.writeSliceWithoutStatistics(slice, 0, length);
}
else {
    directColumnWriter.writeSlice(slice);
}

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/LongDictionaryColumnWriter.java [hunk 10]
```diff
@@ -124,6 +125,9 @@ protected boolean tryConvertRowGroupToDirect(int dictionaryIndexCount, int[] dic
             LongArrayBlock longArrayBlock = new LongArrayBlock(dictionaryIndexCount, Optional.of(directNulls), directValues);
             directColumnWriter.writeBlock(longArrayBlock);
         }
+        if (statistics.isPresent()) {
+            directColumnWriter.finishRowGroup();
+        }
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/LongDictionaryColumnWriter.java:130]: We still calculate the stats if it is present?

**Review by highker — COMMENTED** [review]
- We probably wanna rename `LongColumnWriter` into `LongDirectColumnWriter`
- Regarding the "Reuse String Statistics on Direct Conversion" commit, if you don't mind, I'm leaning towards moving it out of this PR and review it separately.

**Comment by highker** [comment]
Is this ready for review?

**Comment by arunthirupathi** [comment]
My bad, I thought I clicked request review again, but I did not.

**Review by highker — APPROVED** [review]
